wrapper/src/main/java/software/amazon/jdbc/profile/ConfigurationProfile.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.profile;
+
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import software.amazon.jdbc.ConnectionPluginFactory;
+import software.amazon.jdbc.ConnectionProvider;
+import software.amazon.jdbc.authentication.AwsCredentialsProviderHandler;
+import software.amazon.jdbc.dialect.Dialect;
+import software.amazon.jdbc.exceptions.ExceptionHandler;
+import software.amazon.jdbc.targetdriverdialect.TargetDriverDialect;
+
+public class ConfigurationProfile {
+
+  private final @NonNull String name;
+  private final @Nullable List<Class<? extends ConnectionPluginFactory>> pluginFactories;
+  private final @Nullable Properties properties;
+  private @Nullable Supplier<Dialect> dialectSupplier;
+  private @Nullable Supplier<TargetDriverDialect> targetDriverDialectSupplier;
+  private @Nullable Supplier<ExceptionHandler> exceptionHandlerSupplier;
+  private @Nullable Supplier<AwsCredentialsProviderHandler> awsCredentialsProviderHandlerSupplier;
+  private @Nullable Supplier<ConnectionProvider> connectionProviderSupplier;
+
+  private @Nullable Dialect dialect;
+  private @Nullable TargetDriverDialect targetDriverDialect;
+  private @Nullable ExceptionHandler exceptionHandler;
+  private @Nullable AwsCredentialsProviderHandler awsCredentialsProviderHandler;
+  private @Nullable ConnectionProvider connectionProvider;
+
+  private final ReentrantLock lock = new ReentrantLock();
```

Comment:
let's add a comment on why we need a lock on this class, it would be helpful for the future